### ISSUE TITLE
Individual Metrics Features 

### ISSUE THREAD

**Issue by MaddyKapfhammer** [GatorCogitate/cogitate_tool#12]
- How many lines of code an individual added/deleted
- Number of commits per individual
- Number of comments written by an individual (in the code base)
- How many files an individual added/deleted 

(emphasis on first two for demo on 2/18/2020)

**Comment by bagashvilit** [comment]
@noorbuchi I just added demo for Pydriller. Right now the program is looping through the commits and printing how many lines of code the author added/deleted in each file. However, I am guessing that we need to show how many lines of code a person added/deleted in total, instead of in each file. Let me know if you find anything helpful.

**Comment by noorbuchi** [comment]
@bagashvilit thank you, I will check your latest changes and let you know. In your experience so far, did you find Pydriller to more effective and easy to use than PyGithub for this specific feature?

**Comment by bagashvilit** [comment]
I found Pydriller easier to use because here you don't have to generate a personal access token, which, I think, was a downside of PyGithub. Also, I just found out that to calculate some of the metrics Pydriller relies on [lizard](https://github.com/terryyin/lizard). I think its worth to look into that. What do you think of PyGithub?

**Comment by bagashvilit** [comment]
@noorbuchi and I have been discussing how to store the data that we obtain. Please let us know if you have any suggestions about what data structure to use.

**Comment by mcnultycollin** [comment]
@bagashvilit to clarify your question, are you referring to what the return statement should contain? If so, we should use Tuples. Here is a formatting Guidelines from the Software Architecture group:

Key: Tuple ( List , List , List )
where each var represents the following:
GithubID / User : ( Commit List , Issue List , PR List ).

That doesn't quite match with what you're doing right now but in general I would say try to return tuples if possible.

**Comment by mcnultycollin** [comment]
Additionally, @Jordan-A and @WonjoonC what is your status? I would like to move forward with PyGithub unless you have made progress since our last meeting.

**Comment by noorbuchi** [comment]
@mcnultycollin I recently saw what you were referring to. It seems that this is a specific guideline is for storing the data into the .json file. We were referring to the temporary storing of the data in our Python module for processing before writing the complete data set into the .json file. Specifically using either a dictionary or a two-dimensional array. Thank you for pointing out some of the guidelines, we'll keep them in mind.

**Comment by Jordan-A** [comment]
I haven't been able to make progress using Pydriller. I think it would be better to move forward with PyGithub. I just tried running the commit code in our group branch and it won't run locally.

**Comment by bagashvilit** [comment]
@Jordan-A Thanks for the update. I am also working with Pydriller, could you let us know what exactly is going wrong?

**Comment by Jordan-A** [comment]
@bagashvilit  I get a traceback error from my pydriller file in my /python3.8 folder, and a "nosucherror" error.

**Comment by bagashvilit** [comment]
@Jordan-A I just took a look at the `Pipfile` in your branch. One thing I would suggest is to add the necessary package there for Pydriller. Try doing something like this ```[dev-packages]
fastapi = "*"
uvicorn = "*"
black = "*"
pylint = "*"
flake8 = "*"
pydocstyle = "*"
pytest = "*"
PyGithub = "*"
pydriller = "*"
```
I think currently your `Pipfile` is missing `pydriller`.
let me know if this helps.

**Comment by Jordan-A** [comment]
That made PyDriller work for me! I think the problem that I had was specifying the path to the repo. @bagashvilit How do you specify the path to the repository?

**Comment by bagashvilit** [comment]
@Jordan-A Path would be locally on your computer. So something like this `/home/teona/Documents/CS203/labs/cogitate_tool/`

**Comment by Jordan-A** [comment]
Ah I see! I was inputting the Github repository path, which is how it works with PyGithub. That would explain why I got a `no such path error` for PyDriller.

**Comment by Jordan-A** [comment]
@mcnultycollin  What would you like to move forward with in terms of the demos?

**Comment by noorbuchi** [comment]
@Jordan-A I just found out that you can use a URL of a repository to access its contents, however, using a URL would only display statistics for the master branch and not any other active or deleted past branch

**Comment by Jordan-A** [comment]
@noorbuchi Great find! But yes we may want to use an alternative method to allow us to analyze separate branches.

**Comment by noorbuchi** [comment]
@Jordan-A thank you! I've been looking through the Pydriller API and there seems to be a possible way to look into any specific branch but I haven't been able to get it to work. Please let me know if you find anything.

**Comment by mcnultycollin** [comment]
@Jordan-A Lets meet tomorrow, I suspect we'll be able to incorporate both Pydriller and Pygithub. I am feeling pretty good about Pygithub but am open to Pydriller if you are having success with it.

**Comment by Jordan-A** [comment]
@mcnultycollin Sounds great!

**Comment by schultzh** [comment]
Great job collaborating on this issue team! Keep us updated on any solutions you develop!

**Comment by WonjoonC** [comment]
@Jordan-A @mcnultycollin @noorbuchi I think it might be ambitious to say, but it would be great to be able to use a URL of a repository using Pydriller, and somehow possibly travel through each branch using PyGitHub? As Collin stated earlier, I think it would be a good idea to incorporate both tools.

**Comment by mcnultycollin** [comment]
@WonjoonC we can discuss this today, are you free at 4 pm? That is when we were tentatively going to work on our section of the Project.

**Comment by WonjoonC** [comment]
We have been able to incorporate number of commits into the Individual-Metrics-Lines. Our problem that we are facing is duplicates of authors. We are currently looking for ways to solve this issue. Any help would be appreciated.

**Comment by Jordan-A** [comment]
One piece of feedback from the PR suggested that we standardize how we print out the table.  Some of the tools that was suggested to us was tabulate and pretty table. Let's try these tools and decide what direction we want to go.

**Comment by mcnultycollin** [comment]
Hello!
Two things i'm hoping to discuss in our meeting today.
1. I would like to discover the logic behind the total_lines variable in our count_code_lines.py, and whether this has application or not.
2. I believe we should discuss creating a tool to evaluate an Individual metrics score, similar to how group metrics presented.

I met with the Architects group and for this upcoming Tuesday, groups will continue to fetch data using Pydriller independently. Caden will be opening an issue tracker soon to compile a master list of data and features required.

Hope this all made sense, see you guys soon.

**Comment by MaddyKapfhammer** [comment]
Hello Everyone!

Here is rough schedule for our meeting today that I would like to follow!

1. Update on feature progress (Are there any problems? or possible issues?)
2. Review input from the customer from lab session on Tuesday (Thanks @Jordan-A for recording what was said)
3. What would we like to add to our current features or remove from our current features?
4. Would we like to create any new features?
  - Discuss feature for evaluating overall individual score 
  - Discuss feature for determining commit distribution 
5. Assign individuals to specific tasks 
6. What can we get done and when? 
7. Anything that Hannah and I can improve on as project managers?

See you in a few minutes!

**Comment by Jordan-A** [comment]
Meeting transcript: 
1: Small issues?
- Duplication of user names, when you analyze a repository. Pushing from different computers creates new "authors". 
- We can use emails instead of user names.
2: Transcript Review.
- Keep branches open. Working on one branch instead of two.
- Using "Individual-Metrics-Lines"
        - Has commits, lines added/deleted, and total.
Total vs net lines?
- We could change it in the future to changed lines. There's more features in PyDriller relating to this request.
- Fine to have a total lines variable, specifically with PyDriller would be great. Talk with customer
to see if he want's this feature. We don't know the overall evaluation, they may or may not use it.
- If you add 100 lines of code to master then delete 100 lines of code you haven't added anything.
Create tool to evaluate individual only.
- Way to do that without being dragged down by other features.
- Beneficial to evaluate each person individually.
- Let user adjust for the metrics they want to use. That might infringe on the overall evaluation group.
From software architecture group: Everyone is pulling data and therefore we're obsolete.
- Might be good to free up the workload for the overall evaluation group. 

Might be helpful to determine ratio or percentage of number of comments to lines added/deleted. Issue: comments are different depending on commenting language. Python we look for pound sign, Java is different. 
- Prof. Meeting result: Python is very popular. Would be beneficial to create feature for people who use Python. Someone else could maintain the feature that we create. 
- Determine short term features that we want to work with command line interface. Then add more metrics to overall determine how a developer is working in the project. 

3: Add to our current features.
Proposed feature: How about number of lines modified by commit?
- Individually we collect their commits and lines modified.  Lines Modified / Number of Commits. 
- We plan on adding this feature to our demo on Tuesday.

Command line interface issue: Interactive command line interface not ideal for testing and for writing a script. We should not pick the interactive one that was chosen, we want something with a script.
Ex: [Argparse](https://docs.python.org/3/library/argparse.html), pipenv install --dev. Just install the developer packages associated with the repository.

**Comment by bagashvilit** [comment]
I have been researching the difference between the author and the committer on GitHub, and how it can be applied to our project. So basically the author is the person who wrote the code and the committer is the one who actually committed that code on behalf of the author. This ensures that both the author and the committer get credit. For more info please see the [link](https://stackoverflow.com/questions/18750808/difference-between-author-and-committer-in-git). I am not sure if getting the data about the author will fix the issue we are facing.

**Comment by bagashvilit** [comment]
I looked into Pydriller API and instead of obtaining data about `committer`, I got it about `author`. I gave the path for our previous project `Assessment Guidelines` and I think it includes people who merged the pull request too. I would recommend trying both `author = commit.author.name` and `author = commit.committer.name` and see the difference.

**Comment by MaddyKapfhammer** [comment]
Since we are using only one of our branches now, do you think that it is appropriate to close the PR for `Individual-Metrics-Commits`? I think that we should keep the branch open, but as we will not merge it to `master` I think that we should close the currently open PR. Does anyone else have any input on this?

**Comment by munzekm** [comment]
@MaddyKapfhammer I think that is a good idea just so we do not get confused or anyone else gets confused.

**Comment by noorbuchi** [comment]
Hi everyone! I've been working on fixing the duplicate username issue we've been facing and unfortunately, the problem is much more complex than any of us anticipated. Considering that duplication does not occur according to one specific pattern, it is extremely difficult to introduce a fix to every possible reason for duplication. To clarify this, I provided the output of the program I'm working on to demonstrate the wide variety of instances that duplication occurs. Please keep in mind that there could be more instances that we have not encountered yet and thus, we can't account for. 
```
+------------------------+-----------------------------------------------+---------+-----+-----+-------+
|        Username        |                     Email                     | Commits |  +  |  -  | Total |
+------------------------+-----------------------------------------------+---------+-----+-----+-------+
| Gregory M. Kapfhammer  |             gkapfham@allegheny.edu            |    21   | 252 |  42 |  210  |
|    Caden Koscinski     |            koscinskic@allegheny.edu           |    16   | 268 | 186 |   82  |
|       koscinskic       |  42869140+koscinskic@users.noreply.github.com |    1    |  0  |  0  |   0   |
|        clussier        |             lussierc@allegheny.edu            |    13   | 170 |  88 |   82  |
|      Daniel Reid       |              reidd@allegheny.edu              |    7    |  76 |  35 |   41  |
|    MaddyKapfhammer     |           maddy.kapfhammer@gmail.com          |    9    |  98 |  65 |   33  |
|         Teona          |                you@example.com                |    13   |  98 |  73 |   25  |
|     Collin McNulty     |             mcnultyc@allegheny.edu            |    19   | 129 | 109 |   20  |
|     Hannah Schultz     |             schultzh@allegheny.edu            |    9    |  39 |  25 |   14  |
|       StringerJ        |            stringerj@allegheny.edu            |    11   | 117 |  88 |   29  |
|     Jordan Wilson      |     jordanwilson@Jordans-MacBook-Pro.local    |    16   |  74 |  59 |   15  |
| Christopher Stephenson |           stephensonc@allegheny.edu           |    2    |  39 |  0  |   39  |
|       Cory Wiard       |       corywiard@Corys-MacBook-Pro.local       |    15   | 207 | 148 |   59  |
|     Caden Hinckley     |       chinckley@Cadens-MacBook-Pro.local      |    17   | 155 | 141 |   14  |
|      Declan Casey      |          dcasey@decaseyair2015.local          |    12   | 191 | 172 |   19  |
|      megan munzek      |   meganmunzek@megans-MBP.hsd1.pa.comcast.net  |    8    |  38 |  16 |   22  |
|       Tyler Pham       |              phamt@allegheny.edu              |    3    |  26 |  11 |   15  |
|        WonjoonC        |               chos@allegheny.edu              |    8    |  77 |  77 |   0   |
|       wilsonj24        |  31478966+wilsonj24@users.noreply.github.com  |    18   |  0  |  7  |   -7  |
|      Xingbang Liu      |              liux2@allegheny.edu              |    2    |  5  |  5  |   0   |
|      Claire Klima      |              klimac@allegheny.edu             |    1    |  7  |  1  |   6   |
|        Jordan-A        |     jordanbyrne@Jordans-MacBook-Pro.local     |    3    |  24 |  24 |   0   |
|   Christian Lussier    |   32375724+lussierc@users.noreply.github.com  |    5    |  37 |  64 |  -27  |
|        enpuyou         |   35606694+enpuyou@users.noreply.github.com   |    47   |  52 |  51 |   1   |
|    Chris Stephenson    |           stephensonc@allegheny.edu           |    2    |  16 |  5  |   11  |
|       corlettim        |  35552969+corlettim@users.noreply.github.com  |    1    |  0  |  0  |   0   |
|      barrezuetai       | 42876250+barrezuetai@users.noreply.github.com |    2    |  0  |  0  |   0   |
|      bagashvilit       |           bagashvilit@allegheny.edu           |    4    |  13 |  11 |   2   |
|      stephensonc       | 42776587+stephensonc@users.noreply.github.com |    2    |  0  |  0  |   0   |
+------------------------+-----------------------------------------------+---------+-----+-----+-------+
```
This is the output when I provide the link to our conduct repository as the path. Notice some of the instances that duplication occurs.

1. The more general duplication (I plan on fixing that): an email that ends with `@users.noreply.github.com` and starts with some numbers followed by a `+` sign. Notice that the username is between the `+` and `@` sign. I have come up with a solution for this issue that works as long as the username in surrounded by the signs is a key for another set of data. In my solution, I try to detect this issue, merge the data appropriately, then deleted the invalid key containing the invalid email.

2. Issue similar to the previous with the exception that the appropriate key is the first and last name of the user rather than their username making it difficult to detect which data set to merge with. This can be seen in the cases of `Caden Koscinski` conflict with the entry below it `kosinskic`. While both rows belong to the same person, there is very little detectable similarity to merge their data.

3. Issue involving unique duplicate emails. an example of this issue can be seen in  `Teona` entry where the email is `you@example.com ` making it impossible to detect that this entry is a duplicate. I've also encountered similar issues with other users that have duplicate entries with very different emails. I have not been able to solve this issue and I don't expect this to be solved by our demo on Tuesday. 

Proposed solutions: 
1. Look further into the Pydriller API to find a better way to gather user information. This solution seems unlikely to me because Pydriller offers very specific and limited ways to gather information such as usernames and emails. It is worth, however, to look further into this issue.
2. Look into PyGithub API and implement a similar program to see if this issue is also present there. This might be a time-consuming solution but if it is successful, it would allow much more accurate reporting of data despite causing us to refactor our entire module to interact with the PyGitHub API instead of Pydriller.

I hope that you read through this long comment and you have some feedback regarding this issue. Please let me know if you have any suggestions or questions. I realize that this is a difficult issue to explain and I would be happy to clarify anything.

**Comment by WonjoonC** [comment]
@noorbuchi I personally think that tackling what we can for the duplicate problem using Pydriller and getting majority of the duplicates would be good enough with proper documentation explaining this issue. While this would be an annoying "bug" to encounter, I feel that researching PyGitHub and refactoring our program would take too much time and risk. This doesn't necessarily mean to completely throw away the idea, but I would say that getting rid of majority of the duplicates would be suffice for this project.

**Comment by Jordan-A** [comment]
@WonjoonC I'm thinking similar thoughts as @WonjoonC. For this week I think it would make sense to do what we can without refactoring the program. Next week we can chose whether or not to refactor our program to solve this issue. 

As a side not @mcnultycollin and I are working on our task tonight and we'll update you on what we find! 😎

**Comment by MaddyKapfhammer** [comment]
It seems like you all are making great progress, even if it is just in discovering more issues that we are currently having. Thank you for your very detailed description @noorbuchi, I agree with @WonjoonC and @Jordan-A that switching to using `PyGithub` could be dangerous, considering how much progress we have made using `PyDriller`. Perhaps this is an issue that we could discuss with Dr. Kapfhammer in further detail in lab on Tuesday. 

As for the other features and issues that we are having, what is our progress? Are we on track to accomplish what we need to by Tuesday? Is there anything that I can lend a hand with code-wise or in any other way? Thank you all for your hard work this week!

**Comment by mcnultycollin** [comment]
Hello, we are not currently having success pulling lines changed with pydriller. Using file.change_type.name is a way to determine whether lines are being added, deleted or modified, 
but there doesn't seem to be a good way to add this to the dictionary that I'm aware of. 
For some reason, attempting to add if to the dictionary gives me this error
```
Exception("Row has incorrect number of values, (actual) %d!=%d (expected)" %(len(row),len(self._field_names)))
Exception: Row has incorrect number of values, (actual) 9!=8 (expected)
```
Even though I have added the extra index space to the dictionary. Anyone dealt with this sort of thing before?

**Comment by noorbuchi** [comment]
@mcnultycollin this is an issue in regards to the table and not pydriller, you can solve it by adding another item to the list responsible for setting the headers of the table

**Comment by bagashvilit** [comment]
Just a quick note, I created a separate file `print_table.py` in `src`, if you need to call any method and print the data, please refer to that file. I made this modification because `count_code_lines.py` is technically not supposed to be printing out data, and we are just doing it for demonstration purposes, please let me know if anyone has any suggestion

**Comment by Jordan-A** [comment]
Thanks for the advice @noorbuchi ! Great idea to move the print function to a different file @bagashvilit .

**Comment by mcnultycollin** [comment]
@noorbuchi Could you please specify which part of the code you're referring to? I believe I have already attempted this. (We can discuss in class tomorrow)

**Comment by mcnultycollin** [comment]
Hi all, I have just pushed the modify method.

I have found a way to get lines modified for each author. Take a look at this code when you can, this format is built so that other keys can be added easily, such as the counter for added and deleted lines. This is because the dictionary does not have pre-defined keys.
Currently, I have not integrated the method into the print_lines.py and have simply been testing it with a method call at the bottom of count_code_lines.py

**Comment by bagashvilit** [comment]
@noorbuchi I just saw your message in Slack regarding our program that your team will be getting raw data from `.json` file, and process the data to be a dictionary with username as a `key`, how would you store the `values`? Would you use a list or other data structure, this is important for us because we would have to sort that data.

**Comment by noorbuchi** [comment]
@bagashvilit  we will continue to use the current dictionary that we have, we might need to get rid of some unneeded keys such as the dates list and the files list since it will be stored in other modules and it will be specific to each commit. We will retrieve the data that we need from the .json file and calculate totals and other metrics essentially making a dictionary like this: 
```
data[author] = 
{
"EMAIL": email
"COMMITS" : total_commits
"ADDED" : total_added_lines
"REMOVED" : total_removed_lines
"TOTAL" : total_lines (it's just total_added_lines - total_removed_lines)
"MODIFIED" : modified_lines (it's just total_added_lines + total_removed_lines)
"RATIO" : lines_per_commit (it's just modified_lines / total_commits)
"TYPES" : [] (list of file types without duplicates)
}
```
we can add additional data to this structure but this seems like this is the best way to store the data for later on when it needs to be sorted and displayed accordingly. Please let me know what you think and if you have any questions.

**Comment by noorbuchi** [comment]
@bagashvilit I noticed your last commit regarding the sample data, while the outer dictionary structure of this data is correct, the values for the keys shown as the usernames should be other dictionaries rather than lists essentially looking like this.
```
 github_data = {
"noorbuchi" : {"EMAIL": email, "COMMITS" : total_commits, "ADDED" : total_added_lines, "REMOVED" : total_removed_lines},
"bagashvilit" : {"EMAIL": email, "COMMITS" : total_commits, "ADDED" : total_added_lines, "REMOVED" : total_removed_lines},
"Jordan-A" : {"EMAIL": email, "COMMITS" : total_commits, "ADDED" : total_added_lines, "REMOVED" : total_removed_lines}
}
```
I hope this clarifies things a little more, I realize that my previous comment did not mention that the structure should be a dictionary where the value of each key is another dictionary and that could have made things more confusing. Outer keys for the dictionary are usernames, while inner keys are the strings "EMAIL", "COMMITS", "ADDED", and so on depending on what information we want to add.

**Comment by bagashvilit** [comment]
@noorbuchi Thanks again for clarifying, that's exactly what I thought, after I pushed the commit.

**Comment by noorbuchi** [comment]
Hi everyone! I have some updates about the changes that were just implemented into our branch. Some of these changes are significant and they fundamentally change how our feature retrieves and outputs data. The most significant change is the removal of Pydriller when it comes to collecting data. Instead, data will be retrieved using the `data_collection` module already implemented by other groups. Their data will then be written on the .json file as raw unedited data formatted as a list of all the commits in a repository. Our module reads this data from the json file and makes necessary and helpful calculations such as the total number of commits, added lines, removed lines, etc.. for an individual and restructures this data so the username is the key of the dictionary. I encourage all of our team members to take a look to understand this data structure. After some of the metrics have been calculated, they are written back into the json file under a different key. Our `print_table` module runs this same process and iterates through the calculated information in the json file to display the information we want. Please let me know what you think about this new approach, I realize that there are few drawbacks when it comes to testing, but I think it reduces the running time of our tool by not using Pydriller again.

**Comment by Jordan-A** [comment]
@noorbuchi I agree that by not using PyDriller every time we run the program it will be more efficient. I think that the testing tradeoffs are worth it to improve the speed of our tool for now. But we should aim to find ways to get around these drawbacks. I have some thoughts about how we can improve our current system. I think it would be a great idea to have a `main.py` were we can control what functions are run. Maybe for now with a simple command line interface? Ex: `-pd "pydriller"`. Also just to confirm is the `data_processor.py` file intended contain code that reduces the redundancy of authors, etc? Thanks for explaining the changes and great work!

**Comment by noorbuchi** [comment]
@Jordan-A I think you make some great points. We should definitely try to get around these drawbacks and write test cases that make sure our module is working correctly. I also think that a command-line interface would work great, however, other members of our group are working with other groups to get that feature up and running as soon as possible. I would love to see what they make and how we can use it to give more options to the user. In the meantime, I think we should try to come up with some interesting metrics that we can calculate from the data stored in the json files. Additionally, we should start thinking ahead about how we want to incorporate additional data from PyGitHub into the metrics we've established. As for the `data_processor.py` file, I believe that it was made by other groups for their purposes and it became part of our branch when we merged master into our branch. Thank you for your feedback! please let me know what you think about our plan for the upcoming days.

**Comment by Jordan-A** [comment]
@noorbuchi It will be interesting to see what they'll put in the cli for the user! I like the plan that you have set up for the next few days. Thanks for clarifying that the `data_processor.py` file was put into our branch by another team.

**Comment by noorbuchi** [comment]
@Jordan-A I'm happy to clarify! I want to ask all of our team members about the `count_lines_pygithub.py` we currently have in our branch. It is causing multiple style violations and is not being used at all. Do you think that deleting this file is ok, I personally see no use in it and I think it is extra clutter. If everyone agrees, I will remove the file to prevent ant future complications.

**Comment by WonjoonC** [comment]
@noorbuchi I've been following as best as I could remotely to see the progress made. I agree that using Pydriller every time to collect data would be inefficient. I also agree that the PyGitHub file may be obsolete and cause trouble, but it may be difficult to be able to evaluate issues using Pydriller.  How do others think about this issue?

**Comment by noorbuchi** [comment]
@WonjoonC thanks for bringing this up. Considering that our group is only concerned with individual metrics regarding the commits and the additions/deletions for each, I don't think we should focus on tracking interactions just yet. Once we feel confident with what we have after testing and fixing any remaining bugs we can work with other teams to incorporate issues and pull requests into the overall metrics for users. The current PyGithub file in our branch is not complete in any way and it offers no value since we managed to use Pydriller to get the same data. I hope this clarifies my reasoning for why we should get rid of the file. I'm looking forward to hearing from others before removing the file from our branch.

**Comment by Jordan-A** [comment]
@noorbuchi Yes we should complete our work and make sure they're no errors. I think that if the `count_lines_pygithub.py` file offers no value we can delete it. But I think we should reach out to other groups to see if they have any use for `PyGithub`. If they say no lets remove redundant code.

**Comment by noorbuchi** [comment]
@Jordan-A I will check the failed tests you mentioned soon and try to implement a fix. The original intention of the Pygithub file was to compare the usage of Pydriller to Pigithub, so considering that we've already made that decision, I think we should delete it. I don't think that other groups will be able to use it either because it is not very functional and it offers information that we've already managed to get using Pydriller.

**Comment by mcnultycollin** [comment]
Looks like the biggest thing towards getting a passing Travis build is this 
```
tests/test_issue_retrieval.py:24: in <module>
    json_handler.get_dict_from_json_file("contributor_data_template"),
src/json_handler.py:14: in get_dict_from_json_file
    with open(os.path.join(data_path, json_name + ".json"), "r") as json_file:
E   FileNotFoundError: [Errno 2] No such file or directory: '../data/contributor_data_template.json'

```
error that you pointed out @Jordan-A . I see the comment that says it passes with an actual token, maybe we can use a user input that requires a token to run this test suite, or maybe we could set up a dummy Github account with a token that could be hard coded in. I added some comments to describe the code so make sure to pull at some point.

**Comment by noorbuchi** [comment]
@mcnultycollin the issue that you mentioned in the previous comment is referring to the directory of the json file. I'm not sure why you are having this issue considering that this same directory works fine on my device. Considering that we are not using PyGithub, tokens are not needed to access a repository.

**Comment by mcnultycollin** [comment]
I am no longer getting this error.

**Comment by munzekm** [comment]
What did you do in order to get rid of this error? @mcnultycollin

**Comment by mcnultycollin** [comment]
@munzekm TBH i didn't take any steps to fix it. I tried it in various locations but in the end was able to get it to work in root with no debugging. Sorry that this isn't much help.

**Comment by mcnultycollin** [comment]
Hi Team, spoke with the architects and here's what we need to get our pull request merged:

- individual_metrics.py driller.py and print_table.py should be refactored into data_collection.py

- merge_duplicate_usernames.py should be deleted if the issue it is addressing is beyond the scope of our tool.

- Not sure what branch the Individual Metrics evaluation / Dry Erase Board team is in, but that should be factored into data_processing as well as the Team Evaluation.

This way the final tool will only have 4 files in src:
 data_collection, data_processing, json_handler, and cogitate.

I would be happy to assist with any of this, feel free to let me know! Otherwise I will be working on the Readme.

**Comment by Jordan-A** [comment]
@mcnultycollin We have an update! @noorbuchi, @munzekm and I just merged our work, including the refactored code, into master.  Make sure the merge master into your own branches to get all the new files.

**Comment by bagashvilit** [comment]
@MaddyKapfhammer @noorbuchi I just tested out bringing in data from `individual_metrics_testfile.json`, and it worked, which means I could get data for any repository if I wanted to. But the biggest problem now is that the dictionary we are getting has modified file names, we can't use the names for calculations, I was wondering what's the purpose of them, are we just displaying it to the user? Another important point is that so far I am using a dictionary where keys are usernames and the metrics such as `ADDED`,`REMOVED`, however, the format of the dictionary I am getting is slightly different, it seems to have `INDIVIDUAL_METRICS` as a key too. Here is the dictionary I am getting from the json file ```{'INDIVIDUAL_METRICS': {'schultzh': {'EMAIL': '42979565+schultzh@users.noreply.github.com', 'COMMITS': 3, 'ADDED': 60, 'REMOVED': 0, 'TOTAL': 60, 'MODIFIED': 60, 'RATIO': 20, 'FILES': ['README.md', 'travis.yml'], 'FORMAT': ['.md', '.yml']}, 'WonjoonC': {'EMAIL': 'chos@allegheny.edu', 'COMMITS': 4, 'ADDED': 876, 'REMOVED': 874, 'TOTAL': 2, 'MODIFIED': 1750, 'RATIO': 437, 'FILES': ['.flake8', '.gitignore', '.mdlrc', '.pylintrc', '.style.yapf', 'Pipfile', 'README.md', 'demo.py', 'lint.sh', 'test.sh'], 'FORMAT': ['', '.md', '.py', '.sh', '.yapf', 'Pipfile']}, 'Hannah Schultz': {'EMAIL': 'schultzh@allegheny.edu', 'COMMITS': 1, 'ADDED': 0, 'REMOVED': 0, 'TOTAL': 0, 'MODIFIED': 0, 'RATIO': 0, 'FILES': ['.travis.yml'], 'FORMAT': ['.yml']}, 'Jordan-A': {'EMAIL': 'jordanbyrne@Jordans-MacBook-Pro.local', 'COMMITS': 2, 'ADDED': 0, 'REMOVED': 0, 'TOTAL': 0, 'MODIFIED': 0, 'RATIO': 0, 'FILES': ['Commits.py'], 'FORMAT': ['.py']}, 'noorbuchi': {'EMAIL': 'buchin@allegheny.edu', 'COMMITS': 1, 'ADDED': 694, 'REMOVED': 0, 'TOTAL': 694, 'MODIFIED': 694, 'RATIO': 694, 'FILES': ['Pipfile', 'Pipfile.lock', 'UsingPyGithub.py', 'lint.sh', 'test.sh'], 'FORMAT': ['.lock', '.py', '.sh', 'Pipfile']}, 'Chris Stephenson': {'EMAIL': 'stephensonc@allegheny.edu', 'COMMITS': 2, 'ADDED': 89, 'REMOVED': 13, 'TOTAL': 76, 'MODIFIED': 102, 'RATIO': 51, 'FILES': ['demofile.json', 'json_demo.py'], 'FORMAT': ['.json', '.py']}}}
```
Please let me know if you have any opinion

**Comment by noorbuchi** [comment]
@bagashvilit I'm glad your feature is up and running! Please keep in mind that `individual_metrics_testfile.json` should not be changed because it is to be used exclusively by the test suite. As for the modified files metric, the names of all the files a person changed are stored for multiple reasons. When building this function, there was a feature in mind where the user can choose to view the most changed files in a repository. Even though this feature doesn't seem to be the focus anymore, we've decided to keep this metric for future use if needed. I think that it does not cause any trouble and it could be helpful for the overall evaluation if you want to use the number of files modified, which could reflect the knowledge of the tool as a whole rather the focus on one file. This can also be reflected in the types of files changed as well, where you can take the size of that list. I hope this clarifies the intention behind it. You also brought up the dictionary structure stored in the `.json` file. `INDIVIDUAL_METRICS` is key for the dictionary containing the calculated data similar to how `RAW_DATA` is the key for a list of dictionaries for commit hashes. This structure was used to keep both the unedited data and the calculated data in the same json file while being able to distinguish between them. I suggested the use of two `.json` file for the architecture team but they felt that this was better. I hope this answers all of your questions, let me know if there is anything I can help with.

**Comment by bagashvilit** [comment]
@noorbuchi Thank you, I'll consider that, and update you about my progress

**Comment by MaddyKapfhammer** [comment]
Hey guys! I am working on implementing the nested dictionary of the actual data retrieved from GitHub into the `IQR_team_scoring.py` file. I'm running into a lot of difficulty with this currently, would anyone be able to help me with the specific lines of code or commands that I need to run that I should write in order to access the nested dictionary? @noorbuchi @Jordan-A @munzekm I think that you guys may be best suited to help me out with this issue, any help would be appreciated!

**Comment by bagashvilit** [comment]
@MaddyKapfhammer I am getting the data from json did you follow the steps that I send, or exactly what are you having trouble with. Also even if you get it data is in triple nested dictionary when our fake data is different, so we will need to communicate with architects team to fix that

**Comment by MaddyKapfhammer** [comment]
@bagashvilit I did follow the steps that you gave me, and can print the nested dictionary with no problem. I am running into troubles when I try to assign retrieve the dictionary and assign it to a variable in my program. 

This is the error in my terminal window that I am receiving:
```
Traceback (most recent call last):
  File "src/IQR_team_score.py", line 120, in <module>
    print(calculate_average(data_collection.calculate_individual_metrics("individual_metrics_testfile")), "%")
  File "src/IQR_team_score.py", line 108, in calculate_average
    commits_score = calculate_iqr_score(commit_data)
  File "src/IQR_team_score.py", line 51, in calculate_iqr_score
    q1, q3 = np.percentile(dataset, [25, 75])
  File "<__array_function__ internals>", line 6, in percentile
  File "/home/maddykapfhammer/.local/share/virtualenvs/cogitate_tool-HX9JeUoA-/home/maddykapfhammer/.pyenv/shims/python/lib/python3.7/site-packages/numpy/lib/function_base.py", line 3706, in percentile
    a, q, axis, out, overwrite_input, interpolation, keepdims)
  File "/home/maddykapfhammer/.local/share/virtualenvs/cogitate_tool-HX9JeUoA-/home/maddykapfhammer/.pyenv/shims/python/lib/python3.7/site-packages/numpy/lib/function_base.py", line 3826, in _quantile_unchecked
    interpolation=interpolation)
  File "/home/maddykapfhammer/.local/share/virtualenvs/cogitate_tool-HX9JeUoA-/home/maddykapfhammer/.pyenv/shims/python/lib/python3.7/site-packages/numpy/lib/function_base.py", line 3403, in _ureduce
    r = func(a, **kwargs)
  File "/home/maddykapfhammer/.local/share/virtualenvs/cogitate_tool-HX9JeUoA-/home/maddykapfhammer/.pyenv/shims/python/lib/python3.7/site-packages/numpy/lib/function_base.py", line 3941, in _quantile_ureduce_func
    x1 = take(ap, indices_below, axis=axis) * weights_below
  File "<__array_function__ internals>", line 6, in take
  File "/home/maddykapfhammer/.local/share/virtualenvs/cogitate_tool-HX9JeUoA-/home/maddykapfhammer/.pyenv/shims/python/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 194, in take
    return _wrapfunc(a, 'take', indices, axis=axis, out=out, mode=mode)
  File "/home/maddykapfhammer/.local/share/virtualenvs/cogitate_tool-HX9JeUoA-/home/maddykapfhammer/.pyenv/shims/python/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 61, in _wrapfunc
    return bound(*args, **kwds)
IndexError: cannot do a non-empty take from an empty axes.
```

**Comment by MaddyKapfhammer** [comment]
I have since realized that the nested dictionary is being assigned to my variable correctly. I believe that I am getting this error because the dictionary has String attributes when I am trying to do numerical calculations.

**Comment by bagashvilit** [comment]
@MaddyKapfhammer That's definitely one of the reasons also as you see the structure of the dictionary is different, it like triple nested, which also failed my functions. I was also thinking after we are done with our programs why don't we put them in one file, so that we access dictionary from `.json` file only once.

**Comment by noorbuchi** [comment]
@MaddyKapfhammer I hope that you found a way to solve the issue, @bagashvilit and I have been discussing potentially using two `.json` file instead of just one and I think that many people would agree that this would be for the best. The architects' team did not like this idea previously but we should try to talk to them about it again. Also, we've discussed the ways that we will integrate our programs and we came to the conclusion that I should work on the function that merges duplicate usernames and return a dictionary with no calculated data. And your group should do the calculations and write the data to the json file. Please keep in mind that you will get your data from our function and not the json file, I think this would be better to avoid repeatedly writing and erasing data from the file. Let me know if you have any other concerns and please keep us updated with your progress.

**Comment by MaddyKapfhammer** [comment]
@bagashvilit @noorbuchi I also agree that it would be better to use two `.json` files rather than one, we definitely should talk to the architects about this. I think that the idea of merging duplicate usernames and returning a dictionary with no calculated data, so that my group does the calculations and writes to the `.json` file also makes sense. I will let you know what other progress I continue to make tonight, and we should talk to some or all of the architects tomorrow!

**Comment by bagashvilit** [comment]
@MaddyKapfhammer , @noorbuchi I raised an issue on GitHub regarding multiple json files. @MaddyKapfhammer in our branch please use `individual_score_test.json`, the raw data of our branch is there

**Comment by MaddyKapfhammer** [comment]
@bagashvilit I am receiving this error when I use `individual_score_test.json`: `FileNotFoundError: [Errno 2] No such file or directory: './data/individual_score_test.json'` do you have any idea what I can do to fix this error?

**Comment by bagashvilit** [comment]
@MaddyKapfhammer What directory do you run the program from

**Comment by bagashvilit** [comment]
Make sure that you are not in the `src` and run `pipenv run python src/Individual_score.py` but with your file name, let me know if this fixes your issue

**Comment by noorbuchi** [comment]
@bagashvilit can you please clarify the names of the files used to store the raw data and the name of the file for the individual metrics. Additionally, make sure to add 
```
{

}
```
to your json so it is not completely empty

**Comment by bagashvilit** [comment]
@noorbuchi I'll do separate `.json` files for raw data and individual metrics data after I get confirmation from the architects' team. Until then I think I'll keep using one file for my branch.
 I did add the `{ }` to my json file as you recommended before, but thanks for the reminder again.

**Comment by MaddyKapfhammer** [comment]
@bagashvilit I tried the command that you gave me, with both your file and my file, but am still receiving the same error. I noticed that I do not have `individual_score_test.json` in my data folder. Would this be the problem? Do I need to create it, or do you need to push this change?

**Comment by bagashvilit** [comment]
@MaddyKapfhammer I thought I already pushed it, you don't need to create it

**Comment by noorbuchi** [comment]
@bagashvilit I just added the `lines_of_code` metric I told you about today, I'm still not sure what it means but it has some weird and unreasonable number as shown in this comment, please let me know if it's of any use to you or if I should get rid of it.
```
+------------------------+-----------------------------------------------+---------+------+------+---------------+
|        Username        |                     Email                     | Commits |  +   |  -   | Lines of code |
+------------------------+-----------------------------------------------+---------+------+------+---------------+
|        schultzh        |   42979565+schultzh@users.noreply.github.com  |    7    |  60  |  3   |       38      |
|        WonjoonC        |               chos@allegheny.edu              |    5    | 885  | 875  |      580      |
|     Hannah Schultz     |             schultzh@allegheny.edu            |    58   | 1235 | 2051 |      4792     |
|        Jordan-A        |     jordanbyrne@Jordans-MacBook-Pro.local     |    25   | 747  | 583  |      4558     |
|       noorbuchi        |              buchin@allegheny.edu             |    65   | 2821 | 1006 |      7927     |
|    Chris Stephenson    |           stephensonc@allegheny.edu           |    59   | 519  | 346  |      8319     |
|      bagashvilit       |           bagashvilit@allegheny.edu           |    23   | 273  | 350  |      758      |
|     Thomas Cassidy     |  35606317+cassidyt2@users.noreply.github.com  |    2    | 330  | 330  |      272      |
|      Claire Klima      |              klimac@allegheny.edu             |    12   | 570  | 137  |      1111     |
|    MaddyKapfhammer     |           maddy.kapfhammer@gmail.com          |    42   | 2147 | 166  |      8679     |
|    Caden Koscinski     |            koscinskic@allegheny.edu           |    19   | 2378 | 644  |      3901     |
|   Madelyn Kapfhammer   |           maddy.kapfhammer@gmail.com          |    2    |  0   |  0   |       0       |
|      Xingbang Liu      |              liux2@allegheny.edu              |    27   | 1509 | 725  |      3636     |
|      Jordan Byrne      |   46755909+Jordan-A@users.noreply.github.com  |    1    |  1   |  1   |       25      |
|      Pedro Carmo       |         carmop@aldenv169.allegheny.edu        |    18   | 186  | 128  |      516      |
|        johnsc1         |              johnsc@allegheny.edu             |    12   |  61  |  32  |      200      |
|       StringerJ        |            stringerj@allegheny.edu            |    9    | 313  | 1847 |      584      |
| Christopher Stephenson | 42776587+stephensonc@users.noreply.github.com |    9    |  0   |  0   |       0       |
|     devin spitalny     |     devinspitalny@devins-MacBook-Air.local    |    10   |  44  |  33  |      276      |
|      megan munzek      |      meganmunzek@megans-MacBook-Pro.local     |    5    |  10  |  8   |      339      |
|         Caden          |            koscinskic@allegheny.edu           |    9    | 211  | 1358 |      1107     |
|     Collin McNulty     |             mcnultyc@alleheny.edu             |    12   |  74  |  72  |      674      |
|   Marisol Santa Cruz   |  35505388+santacruzm@users.noreply.github.com |    1    | 138  |  0   |      115      |
|         Pedro          |              carmop@allegheny.edu             |    18   | 1943 | 468  |      2062     |
|        clussier        |             lussierc@allegheny.edu            |    4    |  11  |  7   |      160      |
|   Christian Lussier    |   32375724+lussierc@users.noreply.github.com  |    1    |  0   |  0   |       0       |
|     Caden Hinckley     |       chinckley@Cadens-MacBook-Pro.local      |    6    |  9   |  16  |      263      |

```

**Comment by bagashvilit** [comment]
@noorbuchi As you mentioned, we don't really know what that number is doing so I guess we can get rid of it. What do you think?

**Comment by noorbuchi** [comment]
@bagashvilit I'm honestly not sure what it represents, I might keep it for now but if we don't need it, then we can remove it. @MaddyKapfhammer what do you think about this?

**Comment by MaddyKapfhammer** [comment]
@bagashvilit @noorbuchi I am also not sure what this number represents, so I think that it is fine to remove it. It doesn't make much sense to return data that we don't understand the purpose of, the user then won't understand it either.

**Comment by noorbuchi** [comment]
@MaddyKapfhammer Thanks for the feedback, I removed this metric from the uncalculated data. As of right now, the structure of the dictionary is 
```
"INDIVIDUAL_METRICS": {
        "USERNAME": {
            "EMAIL": " ",
            "COMMITS": 0,
            "ADDED": 0,
            "REMOVED": 0,
            "TOTAL": 0,
            "MODIFIED": 0,
            "RATIO": 0,
            "FILES": [" "],
            "FORMAT": [" "],
            "issues_commented": [0],
            "pull_requests_commented": [0],
            "issues_opened": [0],
            "pull_requests_opened": [0]
        }
```
More changes will be made as we discuss the best way to store the data and to help your group access it easily.

**Comment by noorbuchi** [comment]
@MaddyKapfhammer @bagashvilit Hello! As you may know, the structure of the dictionary has been updated with the recent merge into the master branch. The new one will not include the `INDIVIDUAL_METRICS` key, Instead, it will look like I have below. It will also be located in a different json file called `individual_metrics_storage.json`

```
{
"USERNAME1": {
            "EMAIL": " ",
            "COMMITS": 0,
            "ADDED": 0,
            "REMOVED": 0,
            "TOTAL": 0,
            "MODIFIED": 0,
            "RATIO": 0,
            "FILES": [" "],
            "FORMAT": [" "],
            "issues_commented": [0],
            "pull_requests_commented": [0],
            "issues_opened": [0],
            "pull_requests_opened": [0]
        },
"USERNAME2": {
            "EMAIL": " ",
            "COMMITS": 0,
            "ADDED": 0,
            "REMOVED": 0,
            "TOTAL": 0,
            "MODIFIED": 0,
            "RATIO": 0,
            "FILES": [" "],
            "FORMAT": [" "],
            "issues_commented": [0],
            "pull_requests_commented": [0],
            "issues_opened": [0],
            "pull_requests_opened": [0]
        }
}
``` 
More changes will be made as I add keys for commits regarding testing so please keep that in mind, I will keep you updated as those changes are added. Additionally, I have been having issues with the test cases in the `Individual-Metrics-Lines` branch where even though test cases call a specific function, coverage for that funciton is not counting in `codecov`. Any help with that would be appreciated.

**Comment by bagashvilit** [comment]
@noorbuchi @MaddyKapfhammer I moved the calculations for metrics such as `TOTAL`, `RATIO` and so on, and it is working, however, the new metrics that were added such as Github issues, and pull requests are returning empty lists. @noorbuchi Let me know if you have an idea why this is happening.

**Comment by noorbuchi** [comment]
@bagashvilit I'm glad the Pydriller metrics are working. Pygithub metrics, on the other hand, require a different set of functions to be called. I have demonstrated some of that in the main function of the data collection method but I can also put a more detailed explanation here. Please keep in mind that this function is temporary and will not be used in the final version of the too, so I did not try to make it as polished as possible. More error handling still needs to be incorporated and some practices should not be used as I have done
```
DATA = calculate_individual_metrics()
    if DATA == {}:
        print("Raw data was not previously collected, collecting it now...")
        REPO_PATH = input(
            "Enter repository path URL/Local to collect Pydriller data from: "
        )
        collect_and_add_raw_data_to_json(REPO_PATH)
        DATA = calculate_individual_metrics()
```
The previous part is responsible for getting raw data from the REPO_PATH using Pydriller, it also writes the raw data into the appropriate json file depending on the parameters sent in.

```
    # Process for PyGithub data
    token = input("Enter user token to collect PyGitHub data: ")
    repo_name = input("Enter repo name in this format: org/repo_name: ")
    current_repo = authenticate_repository(token, repo_name)
    ISSUE_DATA = {}
    ISSUE_DATA = retrieve_issue_data(current_repo, "all", ISSUE_DATA)
```
The previous section uses the functions written by our arhitects team to get issue data using Pygithub. Please note the parameters needed and refer to the code for additional documentation.

```
    DATA = merge_metric_and_issue_dicts(DATA, ISSUE_DATA)
```
The previous statement merges the two collected dictionaries and sets them equal to data. It also updates the keys in a uniform way through all entries to avoid a key error later on.

```
    print_individual_in_table(data_dict=DATA, headings=["COMMITS"])
```
This is just the printing function

```
    choice = True
    while choice:
        remove = input("enter username to be merged then deleted: ")
        keep = input("enter username to be merged into: ")
        # merge selected entries
        DATA = merge_duplicate_usernames(DATA, keep, remove)
        print("data after this merge...")
        print_individual_in_table(data_dict=DATA, headings=["COMMITS"])
        pick = input("would you like to continue? y/n: ")
        if pick == "n":
            choice = False
```
This loop continuously asks for user input to merge specific users, it still requires some error handling but it is roughly working for the purposes of demoing and testing.

```
    print("Writing data to json file...")
    # Write reformatted dictionary to json, optional parameters not supported
    json_handler.write_dict_to_json_file(DATA, "individual_metrics_storage")
    # print default headings of data
    print_individual_in_table(data_dict=DATA)
```
The final step is writing the data to the json file and printing it out.

I hope this answers your question and gives you an idea of where the issue could be. Please let me know if you have other questions.

**Comment by MaddyKapfhammer** [comment]
With the completion of all of our features, I am going to close this issue. Good work team!
